File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
bool Valid(vector<int>& weights, int days,int mid){
   int total =0,day=1;
  for(int i=0;i<weights.size();i++){
       if(weights[i]> mid) return false;
         if(total + weights[i]<= mid){
            total += weights[i];
         }else{
            day++;
            total = weights[i];
         }
        }

        if(days >= day) return true;

        return false;
  
}
    int shipWithinDays(vector<int>& weights, int days) {
        int sum =0;
        for(int i=0;i<weights.size();i++){
         sum += weights[i];
        }
        int st = 1,end = sum,mid =0,ans =-1;
        while(st<=end){

         mid = st+(end-st)/2;
         if(Valid(weights,days,mid)){
            ans = mid;
            end = mid-1;
         }
         else{
            st = mid+1;
         }


        }
        return ans;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.java
class Solution {
    public boolean helper(int[] nums, int days,int limit){
        int day =1,wt=0;
        for(int i=0;i<nums.length;i++){
          if(wt+nums[i] > limit){
            
            day+=1;
            wt = nums[i];

          }
          else{
            wt += nums[i];
          }

         if(day > days) return false;
        }

        if(day <= days) return true;
        return false;
    }
    public int shipWithinDays(int[] weights, int days) {
         int sum=0,maxi = 0;
        for(int i=0;i<weights.length;i++){
            sum += weights[i];
            maxi = Math.max(maxi,weights[i]);
        }

        int st = maxi,end = sum,ans=0;

        while(st <= end){
          int mid = (end+st)/2;
          if(helper(weights,days,mid)){
            ans = mid;
            end = mid-1;
          }
          else st = mid+1;
        }
        return ans;
    }
}